# Analyzing the severity of car accidents

## 1. Import Libraries

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report
from sklearn.tree import DecisionTreeClassifier
from sklearn import svm
from sklearn.metrics import f1_score
from sklearn.metrics import accuracy_score
from sklearn import preprocessing
from sklearn.linear_model import LogisticRegression
print('imported')

imported


## 2. Loading Data

### Getting the data from csv file from week 1 i.e, Data-Collisions.csv

In [39]:
#@hidden cell
path = "C:\\Users\\pc\\Desktop\\Data-Collisions.csv"
df = pd.read_csv(path)
print("loaded data")

E:\Anaconda\lib\site-packages\IPython\core\interactiveshell.py:3071: DtypeWarning: Columns (33) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


loaded data


In [5]:
df.head()

SEVERITYCODE           X          Y  OBJECTID  INCKEY  COLDETKEY REPORTNO  \
0             2 -122.323148  47.703140         1    1307       1307  3502005   
1             1 -122.347294  47.647172         2   52200      52200  2607959   
2             1 -122.334540  47.607871         3   26700      26700  1482393   
3             1 -122.334803  47.604803         4    1144       1144  3503937   
4             2 -122.306426  47.545739         5   17700      17700  1807429   

    STATUS      ADDRTYPE   INTKEY  ... ROADCOND                LIGHTCOND  \
0  Matched  Intersection  37475.0  ...      Wet                 Daylight   
1  Matched         Block      NaN  ...      Wet  Dark - Street Lights On   
2  Matched         Block      NaN  ...      Dry                 Daylight   
3  Matched         Block      NaN  ...      Dry                 Daylight   
4  Matched  Intersection  34387.0  ...      Wet                 Daylight   

  PEDROWNOTGRNT  SDOTCOLNUM SPEEDING ST_COLCODE  \
0           NaN         NaN      NaN         10   
1           NaN   6354039.0      NaN         11   
2           NaN   4323031.0      NaN         32   
3           NaN         NaN      NaN         23   
4           NaN   4028032.0      NaN         10   

                                          ST_COLDESC  SEGLANEKEY  \
0                                  Entering at angle           0   
1  From same direction - both going straight - bo...           0   
2                             One parked--one moving           0   
3                   From same direction - all others           0   
4                                  Entering at angle           0   

   CROSSWALKKEY  HITPARKEDCAR  
0             0             N  
1             0             N  
2             0             N  
3             0             N  
4             0             N  

[5 rows x 38 columns]

In [7]:
df.shape

(194673, 38)

### Getting the data types of the dataframe

In [8]:
df.dtypes

SEVERITYCODE        int64
X                 float64
Y                 float64
OBJECTID            int64
INCKEY              int64
COLDETKEY           int64
REPORTNO           object
STATUS             object
ADDRTYPE           object
INTKEY            float64
LOCATION           object
EXCEPTRSNCODE      object
EXCEPTRSNDESC      object
SEVERITYCODE.1      int64
SEVERITYDESC       object
COLLISIONTYPE      object
PERSONCOUNT         int64
PEDCOUNT            int64
PEDCYLCOUNT         int64
VEHCOUNT            int64
INCDATE            object
INCDTTM            object
JUNCTIONTYPE       object
SDOT_COLCODE        int64
SDOT_COLDESC       object
INATTENTIONIND     object
UNDERINFL          object
WEATHER            object
ROADCOND           object
LIGHTCOND          object
PEDROWNOTGRNT      object
SDOTCOLNUM        float64
SPEEDING           object
ST_COLCODE         object
ST_COLDESC         object
SEGLANEKEY          int64
CROSSWALKKEY        int64
HITPARKEDCAR       object
dtype: objec

### Getting an Summary info of the dataframe

In [9]:
df.describe()

SEVERITYCODE              X              Y       OBJECTID  \
count  194673.000000  189339.000000  189339.000000  194673.000000   
mean        1.298901    -122.330518      47.619543  108479.364930   
std         0.457778       0.029976       0.056157   62649.722558   
min         1.000000    -122.419091      47.495573       1.000000   
25%         1.000000    -122.348673      47.575956   54267.000000   
50%         1.000000    -122.330224      47.615369  106912.000000   
75%         2.000000    -122.311937      47.663664  162272.000000   
max         2.000000    -122.238949      47.734142  219547.000000   

              INCKEY      COLDETKEY         INTKEY  SEVERITYCODE.1  \
count  194673.000000  194673.000000   65070.000000   194673.000000   
mean   141091.456350  141298.811381   37558.450576        1.298901   
std     86634.402737   86986.542110   51745.990273        0.457778   
min      1001.000000    1001.000000   23807.000000        1.000000   
25%     70383.000000   70383.000000   28667.000000        1.000000   
50%    123363.000000  123363.000000   29973.000000        1.000000   
75%    203319.000000  203459.000000   33973.000000        2.000000   
max    331454.000000  332954.000000  757580.000000        2.000000   

         PERSONCOUNT       PEDCOUNT    PEDCYLCOUNT       VEHCOUNT  \
count  194673.000000  194673.000000  194673.000000  194673.000000   
mean        2.444427       0.037139       0.028391       1.920780   
std         1.345929       0.198150       0.167413       0.631047   
min         0.000000       0.000000       0.000000       0.000000   
25%         2.000000       0.000000       0.000000       2.000000   
50%         2.000000       0.000000       0.000000       2.000000   
75%         3.000000       0.000000       0.000000       2.000000   
max        81.000000       6.000000       2.000000      12.000000   

        SDOT_COLCODE    SDOTCOLNUM     SEGLANEKEY  CROSSWALKKEY  
count  194673.000000  1.149360e+05  194673.000000  1.946730e+05  
mean       13.867768  7.972521e+06     269.401114  9.782452e+03  
std         6.868755  2.553533e+06    3315.776055  7.226926e+04  
min         0.000000  1.007024e+06       0.000000  0.000000e+00  
25%        11.000000  6.040015e+06       0.000000  0.000000e+00  
50%        13.000000  8.023022e+06       0.000000  0.000000e+00  
75%        14.000000  1.015501e+07       0.000000  0.000000e+00  
max        69.000000  1.307202e+07  525241.000000  5.239700e+06

## 3. Choosing the columns 'SEVERITYCODE' , 'SPEEDING' , 'ROADCOND' data to work with

In [10]:
# choosing the data we will work with
test = ['SEVERITYCODE', 'SPEEDING','ROADCOND']
df = df[test]

# obtaining data dimensions
for feature in ["SPEEDING", "ROADCOND"]:
    print(df[feature].unique())

[nan 'Y']
['Wet' 'Dry' nan 'Unknown' 'Snow/Slush' 'Ice' 'Other' 'Sand/Mud/Dirt'
 'Standing Water' 'Oil']


### Replacing 'nan' values in both 'SPEEDING' & 'ROADCOND' with 'N' & 'Unknown' respectively 

In [11]:
# in speed we replace Nan with a negative value N
df['SPEEDING'] = df['SPEEDING'].fillna('N')


#we replace the value Nan declaring it as unknown too

df['ROADCOND'] = df['ROADCOND'].fillna('Unknown')

# checking value once again...
for feature in ["SPEEDING", "ROADCOND"]:
    print(df[feature].unique())

['N' 'Y']
['Wet' 'Dry' 'Unknown' 'Snow/Slush' 'Ice' 'Other' 'Sand/Mud/Dirt'
 'Standing Water' 'Oil']


### Assigning new values to 'ROADCOND' column

In [12]:
# We assign new values to roadcond
df['ROADCOND'].replace(to_replace=['Wet','Dry','Unknown','Snow/Slush','Ice','Other','Sand/Mud/Dirt','Standing Water','Oil'], value = ['Dangerous','Normal','Normal','Dangerous','Dangerous','Normal','Dangerous','Dangerous','Dangerous'], inplace=True)

In [27]:
df["SPEEDING"] = df.SPEEDING.astype(str)

In [28]:
df["ROADCOND"] = df.ROADCOND.astype(str)

In [29]:
df["SPEEDING"].replace(to_replace=['N', 'Y'], value=[0,1], inplace=True)

In [30]:
df['ROADCOND'].replace(to_replace=['Dangerous','Normal'],value=[0,1],inplace=True)
test_condition = df[['SPEEDING','ROADCOND']]
test_condition.head()

SPEEDING ROADCOND
0        0        0
1        0        0
2        0        1
3        0        1
4        0        0

In [32]:
speed_analysis = df.groupby(['SPEEDING'])['SEVERITYCODE'].value_counts(normalize=True)
speed_analysis

SPEEDING  SEVERITYCODE
0         1               0.705099
          2               0.294901
1         1               0.621665
          2               0.378335
Name: SEVERITYCODE, dtype: float64

## 4.Training the Model

In [33]:
x = test_condition
y = df['SEVERITYCODE'].values.astype(str)
x = preprocessing.StandardScaler().fit(x).transform(x)
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=1234)

# obtaining data dimensions
print("Training set: ", x_train.shape, y_train.shape)
print("Testing set: ", x_test.shape, y_test.shape)

Training set:  (155738, 2) (155738,)
Testing set:  (38935, 2) (38935,)


## 5. Selecting the methods: Tree model, Logistic Regression and KNN methodology

In [34]:
#Tree model
Tree_model = DecisionTreeClassifier(criterion="entropy", max_depth = 4)
Tree_model.fit(x_train, y_train)
predicted = Tree_model.predict(x_test)
Tree_f1 = f1_score(y_test, predicted, average='weighted')
Tree_acc = accuracy_score(y_test, predicted)

In [35]:
#Logistic Regression
LR_model = LogisticRegression(C=0.01, solver='liblinear').fit(x_train, y_train)
predicted = LR_model.predict(x_test)
LR_f1 = f1_score(y_test, predicted, average='weighted')
LR_acc = accuracy_score(y_test, predicted)

In [36]:
#KNN methodology
KNN_model = KNeighborsClassifier(n_neighbors = 4).fit(x_train, y_train)
predicted = KNN_model.predict(x_test)
KNN_f1 = f1_score(y_test, predicted, average='weighted')
KNN_acc = accuracy_score(y_test, predicted)

## 6. Results
### Comparing the results

In [37]:
results = {
    "Method of Analisys": ["KNN", "Decision Tree", "LogisticRegression"],
    "F1-score": [KNN_f1, Tree_f1, LR_f1],
    "Accuracy": [KNN_acc, Tree_acc, LR_acc]
}

results = pd.DataFrame(results)
results

Method of Analisys  F1-score  Accuracy
0                 KNN  0.591378  0.696751
1       Decision Tree  0.576051  0.699679
2  LogisticRegression  0.576051  0.699679

In [38]:
# Comparing results using LR
results = {
    "Intercept": LR_model.intercept_,
    "SPEEDING ": LR_model.coef_[:,0],
    "ROADCOND ": LR_model.coef_[:,1],
}

results = pd.DataFrame(results)
results

Intercept  SPEEDING   ROADCOND 
0  -0.853729   0.067702  -0.068295

# What i have done so far?

### Introduction:

 Traffic accidents represent one of the leading causes of death worldwide and of economic expenditure. Despite the numerous measures and campaigns that are deployed every year to raise awareness of the seriousness of the problem, it still occurs quite frequently. The impact of road accidents on society and the economy is high, and human losses are compounded by large expenditures on health care, awareness campaigns, mobilization of specialized personnel, etc. The WHO sets the economic impact of road accidents in a developed country at 2 to 3% of GDP, a significant figure for any country. Collaboration to reduce these losses has become an important issue of general interest.

* Defining the problem:

 What are the factors that have a high impact on road accidents?

 Is there a pattern to them?

 Correlation?

We will have to analyze the data to get a clearer picture and draw conclusions.

Note that this work represents the final project of the IBM certification course, for which we have provided the data with which we will develop the project.

These data have been collected and shared by the Seattle Police Department (Traffic Records) and are provided by Coursera for downloading through a link.

It takes into account a period of time from 2004 to the present, recording information related to the severity of the traffic accident, location, type of collision, weather and road conditions, visibility, number of people involved, etc.

The objective is to define the problem, to find the factors that can have a relevant weight in the quantity and seriousness of the accidents, so that any organism, company or enterprise interested in reducing these figures, can focus the resources in points where these conditions converge.

In order to provide greater clarity, I will try to analyze the data, see if there are relationships or patterns, especially in high impact accidents, so that preventive measures can focus on these points as a first prevention strategy.



## DATA WE NEED:

For an accurate prediction of the magnitude of damage caused by accidents, they require a large number of reports on traffic accidents with accurate data to train prediction models. The data set provided for this work allows the analysis of a record of 200,000 accidents in the state of Seattle, from 2004 to the date it is issued, in which 37 attributes or variables are recorded and the codification of the type of accident is allowed, grouped according to 84 codes. The information can be extracted from it:

1. speed information
2. information on road conditions and visibility
3. type of collision
4.  affected persons, etc
The data will be used so that we can determine which attributes are most common in traffic accidents in order to target 
prevention at these high-incidence points.
 
 
 * Data Source:
 
   Data Source: These data have been collected and shared by the Seattle Police Department (Traffic Records) and are provided by Coursera for downloading through a link.

    Data Location: Coursera_Capstone/Data assets

    Data set name: Data-Collisions (1)_shaped.csv



## Methodology
* Objective:
 The objective of this project is to predict the severity of a traffic accident based on the other characteristics contained in the report.

* Packages and libraries: 
 We will use libraries and packages for both data manipulation and data visualization. PANDA, NUMPY, SCIPY, Matplotlib, Seaborn

A data analysis will be performed in order to determine what type of methodology and learning of the machine will be the most appropriate, in addition to obtaining a first contact with the data that we find more relevant to use in this project.




## DATA CLEANING:

1. Firstly i imported libraries that are used for this project.
2. Got the data from the csv file to the data frame df given in week 1 i.e,Data-Collisions.csv.
3. found the top 5 rows of the dat frame.
4. Got the datatypes of each column.
5. Got the Summary information of the dataframe using describe().
6. I choosed the columns 'SEVERITYCODE' , 'SPEEDING' , 'ROADCOND' from the data frame df to work with.
7. Replaced 'nan' values in both 'SPEEDING' & 'ROADCOND' with 'N' & 'Unknown' respectively.
8. I assigned new values to the column 'ROADCOND' as 'Dangerous' or 'Normal'.
9. I casted the columns 'SPEEDING' & 'ROADCOND' as type string datatype.
10. I replaced the values of columns 'SPEEDING' with 0 and 1.
11. I had done the same thing with the column 'ROADCOND' i.e, replacing the values.
12. Now i took these two columns 'SPEEDING' &'ROADCOND' from the dataframe to "test_condition".
13. I took a dataframe "speed_analysis" to which i passed by grouping the columns in dataframe 'SPEEDING' & 'SEVERITYCODE',And printed the dataframe.
 


##  DATA ANALYSIS:

1. This Dataset is now Ready for any Machine Learning Algorithm. 
2. I trained the model.
3. Selected the methods: Tree model, Logistic Regression and KNN methodology.
4. Obtained the f1-score and accuracy of the 3 models mentioned above. 
5. I compared the results by LR


## Results:
    Intercept 	SPEEDING	ROADCOND
	-0.853729	0.067702	-0.068295

According to the results of this method, the two attributes of the table have a determining influence on the increase of accidents, however more factors must be taken into account

## Conclusion and recommendations

Knowing that speed and road conditions are factors that influence traffic accidents and their severity, recommendations for reducing accidents could include

* Limiting speed in areas where road conditions are worse

* Surveillance and radar cameras -Maintain traffic control routes on these roads

* Enable emergency exit areas and rescue stations

* Improve the maintenance of these streets, signaling and lighting

## THANK YOU!